File: 2024/day18/day18.go
//! [Day 18: RAM Run](https://adventofcode.com/2024/day/18)

package main

import (
	"bufio"
	"fmt"
	"os"
)

type Point struct {
	x, y int
}

type QueueItem struct {
	pos   Point
	steps uint
}

func readInput(filename string) []Point {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var points []Point
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		var x, y int
		fmt.Sscanf(scanner.Text(), "%d,%d", &x, &y)
		points = append(points, Point{x, y})
	}
	return points
}

func findPath(memory [][]bool, memSize int) uint {

	start := Point{0, 0}
	end := Point{memSize - 1, memSize - 1}

	if memory[start.y][start.x] || memory[end.y][end.x] {
		return 0
	}

	seen := make([][]bool, len(memory))
	for i := range seen {
		seen[i] = make([]bool, len(memory[0]))
	}

	dirs := []Point{{0, 1}, {1, 0}, {0, -1}, {-1, 0}}
	queue := []QueueItem{{start, 0}}
	seen[start.y][start.x] = true

	for len(queue) > 0 {
		pos := queue[0]
		queue = queue[1:]

		if pos.pos == end {
			return pos.steps
		}

		for _, dir := range dirs {
			new_pos := Point{pos.pos.x + dir.x, pos.pos.y + dir.y}
			if new_pos.x >= 0 && new_pos.x < memSize && new_pos.y >= 0 && new_pos.y < memSize &&
				!memory[new_pos.y][new_pos.x] && !seen[new_pos.y][new_pos.x] {
				seen[new_pos.y][new_pos.x] = true
				queue = append(queue, QueueItem{new_pos, pos.steps + 1})
			}
		}
	}
	return 0
}

func main() {
	inputFile := "input.txt"
	if len(os.Args) >= 2 {
		inputFile = os.Args[1]
	}

	byte_positions := readInput(inputFile)
	memSize := 71
	numCorruptions := 1024

	// Part 1
	memory := make([][]bool, memSize)
	for i := range memory {
		memory[i] = make([]bool, memSize)
	}

	for i := 0; i < numCorruptions && i < len(byte_positions); i++ {
		p := byte_positions[i]
		memory[p.y][p.x] = true
	}

	steps := findPath(memory, memSize)
	fmt.Println(steps)

	// Part 2
	memory = make([][]bool, memSize)
	for i := range memory {
		memory[i] = make([]bool, memSize)
	}

	for _, p := range byte_positions {
		memory[p.y][p.x] = true
		steps := findPath(memory, memSize)
		if steps == 0 {
			fmt.Printf("%d,%d\n", p.x, p.y)
			break
		}
	}
}

File: 2024/day18/day18.rs
//! [Day 18: RAM Run](https://adventofcode.com/2024/day/18)

use std::collections::VecDeque;

use aoc::{Coord, Grid};

const CORRUPTED: u8 = 0xCC;

struct Puzzle {
    byte_positions: Vec<Coord>,
    mem_size: i32,
    num_corruptions: usize,
}

impl Puzzle {
    fn new(data: &str) -> Self {
        let mut byte_positions = Vec::new();

        for line in data.lines() {
            let (x, y) = line.split_once(',').unwrap();

            let x: i32 = x.parse().unwrap();
            let y: i32 = y.parse().unwrap();

            byte_positions.push(Coord { x, y });
        }

        Self {
            byte_positions,
            mem_size: 71,
            num_corruptions: 1024,
        }
    }

    fn find_path(&self, memory: &Grid<u8>) -> u32 {
        let mut queue = VecDeque::new();
        let mut seen = Grid::<bool>::with_size(self.mem_size, self.mem_size, false, false);
        // nota: direct access is much faster than using a hashset

        let start = Coord::new(0, 0);
        let end = Coord::new(self.mem_size - 1, self.mem_size - 1);

        queue.push_back((start, 0));
        seen[start] = true;

        while let Some((pos, steps)) = queue.pop_front() {
            if pos == end {
                return steps;
            }

            for (_, new_pos) in memory.iter_directions(pos) {
                if memory[new_pos] != CORRUPTED && !seen[new_pos] {
                    queue.push_back((new_pos, steps + 1));
                    seen[new_pos] = true;
                }
            }
        }

        0
    }

    /// Solve part one.
    fn part1(&self) -> u32 {
        let mut memory = Grid::<u8>::with_size(self.mem_size, self.mem_size, b' ', b'#');

        self.byte_positions
            .iter()
            .take(self.num_corruptions)
            .for_each(|&pos| memory[pos] = CORRUPTED);

        self.find_path(&memory)
    }

    /// Solve part two.
    fn part2(&self) -> String {
        let mut a = 0;
        let mut b = self.byte_positions.len() - 1;

        while a + 1 < b {
            let mut memory = Grid::<u8>::with_size(self.mem_size, self.mem_size, b' ', b'#');

            let m = (a + b) / 2;

            // corrupt the first m bytes
            for &pos in self.byte_positions.iter().take(m) {
                memory[pos] = CORRUPTED;
            }

            if self.find_path(&memory) == 0 {
                b = m;
            } else {
                a = m;
            }
        }

        format!("{},{}", self.byte_positions[a].x, self.byte_positions[a].y)
    }
}

/// # Panics
/// over malformed input
#[must_use]
pub fn solve(data: &str) -> (u32, String) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const SAMPLE_1: &str = include_str!("sample_1.txt");

    #[test]
    fn test01() {
        let mut puzzle = Puzzle::new(SAMPLE_1);
        puzzle.num_corruptions = 12;
        puzzle.mem_size = 7;
        assert_eq!(puzzle.part1(), 22);
    }

    #[test]
    fn test02() {
        let mut puzzle = Puzzle::new(SAMPLE_1);
        puzzle.mem_size = 7;
        assert_eq!(puzzle.part2(), "6,1");
    }
}
